kong/observability/logs.lua

```diff
@@ -0,0 +1,169 @@
+local _M = {
+  maybe_push = function() end,
+  get_request_logs = function() return {} end,
+  get_worker_logs = function() return {} end,
+}
+
+if ngx.config.subsystem ~= "http" then
+  return _M
+end
+
+
+local request_id_get = require "kong.observability.tracing.request_id".get
+local time_ns = require "kong.tools.time".time_ns
+local cycle_aware_deep_copy = require "kong.tools.utils".cycle_aware_deep_copy
+
+local get_log_level = require "resty.kong.log".get_log_level
+local constants_log_levels = require "kong.constants".LOG_LEVELS
+
+local table_new = require "table.new"
+local table_clear = require "table.clear"
+local string_buffer = require "string.buffer"
+
+local ngx = ngx
+local kong = kong
+local table = table
+local tostring = tostring
+
+local MAX_WORKER_LOGS = 1000
+local MAX_REQUEST_LOGS = 1000
+local NGX_CTX_REQUEST_LOGS_KEY = "o11y_logs_request_scoped"
+
+local worker_logs = table_new(100, 0)
+local logline_buf = string_buffer.new()
+
+
+local function configured_log_level()
+  local ok, level = pcall(get_log_level)
+  if not ok then
+    level = constants_log_levels[kong.configuration.log_level]
+  end
+
+  return level
+end
+
+
+-- needed because table.concat doesn't like booleans
+local function concat_tostring(tab)
+  if #tab == 0 then
+    return ""
+  end
+
+  for i = 1, #tab do
+    logline_buf:put(tostring(tab[i]))
+  end
+
+  local retstr = logline_buf:tostring()
+  logline_buf:reset()
+
+  return retstr
+end
+
+
+local function generate_log_entry(request_scoped, log_level, log_str, request_id, debug_info)
+
+  local span_id
+
+  if request_scoped then
+    -- add tracing information if tracing is enabled
+    local active_span = kong and kong.tracing and kong.tracing.active_span()
+    if active_span then
+      span_id = active_span.span_id
+    end
+  end
+
+  local attributes = {
+    request_id = request_id,
+    introspection_current_line = debug_info.currentline,
+    introspection_name = debug_info.name,
+    introspection_namewhat = debug_info.namewhat,
+    introspection_source = debug_info.source,
+    introspection_what = debug_info.what,
+  }
+
+  local now_ns = time_ns()
+  return {
+    time_unix_nano = now_ns,
+    observed_time_unix_nano = now_ns,
+    log_level = log_level,
+    body = log_str,
+    attributes = attributes,
+    span_id = span_id,
+  }
+end
+
+
+local function get_request_log_buffer()
+  local log_buffer = ngx.ctx[NGX_CTX_REQUEST_LOGS_KEY]
+  if not log_buffer then
+    log_buffer = table_new(10, 0)
+    ngx.ctx[NGX_CTX_REQUEST_LOGS_KEY] = log_buffer
+  end
+  return log_buffer
+end
+
+
+function _M.maybe_push(stack_level, log_level, ...)
+  -- !WARNING! no logging here, to avoid infinite recursion.
```

Comment:
It is hard to make sure there is no function call to `ngx.log` in the functions that are called by this function and its calls. If a sub-function call invokes it, we will overflow our stack, which has huge impact on production.

Is it possible to set a flag for detecting the infinite recursion at the early stage, this would have less performance impact rather than stackoverflow.

Reply:
I thought about that as well, I'll give it a try, thanks.

Reply:
I wanted to try a smart solution here, like checking the stack, but anything I came up with was too inefficient. Ended up adding a simple flag and some comment to warn against introducing yielding behavior in the handler called by the hook.